File: src/main/java/org/soraworld/violet/yaml/IYamlConfiguration.java
package org.soraworld.violet.yaml;

import com.google.common.base.Charsets;
import com.google.common.io.Files;
import org.apache.commons.lang.Validate;
import org.bukkit.configuration.InvalidConfigurationException;
import org.bukkit.configuration.file.YamlConfiguration;
import org.bukkit.configuration.file.YamlConstructor;
import org.bukkit.configuration.file.YamlRepresenter;
import org.yaml.snakeyaml.DumperOptions;
import org.yaml.snakeyaml.representer.Representer;

import java.io.*;

public class IYamlConfiguration extends YamlConfiguration {

    private final DumperOptions dumperOptions = new DumperOptions();
    private final Representer yamlRepresent = new YamlRepresenter();
    private final IYaml yaml = new IYaml(new YamlConstructor(), yamlRepresent, dumperOptions);

    @Override
    public String saveToString() {
        dumperOptions.setIndent(options().indent());
        dumperOptions.setDefaultFlowStyle(DumperOptions.FlowStyle.BLOCK);
        dumperOptions.setAllowUnicode(true);
        yamlRepresent.setDefaultFlowStyle(DumperOptions.FlowStyle.BLOCK);

        String header = buildHeader();
        String dump = yaml.dump(getValues(false));

        if (dump.equals(BLANK_CONFIG)) {
            dump = "";
        }

        return header + dump;
    }

    public void load(File file) throws IOException, InvalidConfigurationException {
        Validate.notNull(file, "File cannot be null");
        final FileInputStream stream = new FileInputStream(file);
        load(new InputStreamReader(stream, Charsets.UTF_8));
    }

    public void load(InputStream stream) throws IOException, InvalidConfigurationException {
        Validate.notNull(stream, "Stream cannot be null");
        load(new InputStreamReader(stream, Charsets.UTF_8));
    }

    public void load(Reader reader) throws IOException, InvalidConfigurationException {

        StringBuilder builder = new StringBuilder();

        try (BufferedReader input = reader instanceof BufferedReader ? (BufferedReader) reader : new BufferedReader(reader)) {
            String line;
            while ((line = input.readLine()) != null) {
                builder.append(line);
                builder.append('\n');
            }
        }

        loadFromString(builder.toString());
    }

    public void save(File file) throws IOException {
        Validate.notNull(file, "File cannot be null");
        Files.createParentDirs(file);
        String data = saveToString();
        try (Writer writer = new OutputStreamWriter(new FileOutputStream(file), Charsets.UTF_8)) {
            writer.write(data);
        }
    }

    public void clear() {
        map.clear();
    }

}

File: src/main/kotlin/org/soraworld/violet/yaml/IYamlConfiguration.kt
package org.soraworld.violet.yaml

import org.bukkit.configuration.file.YamlConfiguration
import org.bukkit.configuration.file.YamlConstructor
import org.bukkit.configuration.file.YamlRepresenter
import org.yaml.snakeyaml.DumperOptions
import java.io.*

class IYamlConfiguration : YamlConfiguration() {

    private val dumperOptions = DumperOptions()
    private val yamlRepresent = YamlRepresenter()
    private val yaml: IYaml = IYaml(YamlConstructor(), yamlRepresent, dumperOptions)

    override fun saveToString(): String {
        dumperOptions.indent = options().indent()
        dumperOptions.defaultFlowStyle = DumperOptions.FlowStyle.BLOCK
        dumperOptions.isAllowUnicode = true
        yamlRepresent.defaultFlowStyle = DumperOptions.FlowStyle.BLOCK
        val header = buildHeader()
        var dump = yaml.dump(getValues(false))
        if (dump == BLANK_CONFIG) dump = ""
        return header + dump
    }

    @Throws(Exception::class)
    override fun load(file: File) = load(InputStreamReader(FileInputStream(file), Charsets.UTF_8))

    @Throws(Exception::class)
    override fun load(stream: InputStream) = load(InputStreamReader(stream, Charsets.UTF_8))

    @Throws(Exception::class)
    override fun load(reader: Reader) {
        val builder = StringBuilder()
        (reader as? BufferedReader ?: BufferedReader(reader)).use { input ->
            var line: String? = input.readLine()
            while (line != null) {
                builder.append(line)
                builder.append('\n')
                line = input.readLine()
            }
        }
        loadFromString(builder.toString())
    }

    @Throws(Exception::class)
    override fun save(file: File) {
        val parent = file.canonicalFile.parentFile ?: return
        parent.mkdirs()
        if (!parent.isDirectory) throw IOException("Unable to create parent directories of $file")
        val data = saveToString()
        OutputStreamWriter(FileOutputStream(file), Charsets.UTF_8).use { writer -> writer.write(data) }
    }

    fun clear() {
        map.clear()
    }

}
